src/c2s/mongoose_c2s.erl

```diff
@@ -924,3 +927,11 @@ merge_states(S0 = #c2s_data{},
       jid = S0#c2s_data.jid,
       state_mod = S0#c2s_data.state_mod
      }.
+
+-spec terminate_session(jid:jid() | pid() | none, binary()) -> {exit, binary()} | no_session.
+terminate_session(none, _Reason) ->
+    no_session;
+terminate_session(#jid{} = Jid, Reason) ->
+    terminate_session(ejabberd_sm:get_session_pid(Jid), Reason);
+terminate_session(Pid, Reason) ->
+    Pid ! {exit, Reason}.
```

Comment:
This one is a bit of an artefact from the old `ejabberd_c2s` that we can cleanup here, after all the goal is to have a very nice `mongoose_c2s` code this time 😃 

I'd say that first the API for c2s should be that of a `gen_statem`, simply taking the pid and doing `gen_statem:cast/2`. It is the session manager, `ejabberd_sm`, the one responsible for finding the pid given a jid, if there's any. So this function should be in `ejabberd_sm` instead of here, and then call a new `mongoose_c2s:exit/2` function.

Then the second point is that this new `mongoose_c2s:exit/2` is kinda surprisingly similar to the already implemented `mongoose_c2s:stop/2` one. The thing is that one tags the reason with the atom `stop`, and the other with the atom `exit`. As far as I remember from what I wrote back then, is that `stop` will run a hook that gives a chance to trap the exit (see `handle_stop_request/3`, while `exit` won't. Keep that in mind when we see other modules doing one call or the other, it'd be good to think carefully what is it that the call really wants to do.

But for what we can do here: https://app.codecov.io/gh/esl/MongooseIM/blob/feature%2Fmongoose_c2s/src/c2s/mongoose_c2s.erl my `handle_stop_request/3` is not even used at the moment 😅 And it won't, because it's handling `info` messages, and not casts. So we need a `handle_cast` in the same place we have the `handle_info` (https://github.com/esl/MongooseIM/blob/458e76557b122c644e3b6d7f69e5aa705cc37262/src/c2s/mongoose_c2s.erl#L143), and then handle the two casts we have implemented here so far.